### ISSUE TITLE
Must be able to detect chaining cycle

### ISSUE THREAD

**Issue by amasad** [then/promise#112]
Both native promises in Chrome and Bluebird has that.

![screenshot 2015-10-01 16 29 58](https://cloud.githubusercontent.com/assets/587518/10236394/d9631582-6859-11e5-9cb9-acb74e60988b.png)

**Comment by amasad** [comment]
Actually, nvm, you do detect it (it's just there is no uncaught error warning). My issue turns out is that I'm calling `then` before I return the chained promise:

```
var b = Promise.resolve().then(function() {return b.then(function() {return 1;})});
```

This obviously, never finishes, yet it's not detected.

**Comment by ForbesLindesay** [comment]
Indeed, your original example is covered by the Promises/A+ test suite, so we definitely detect that.

The later example is (as discussed on twitter) vastly more complex to actually detect.  If you think it's worth while, we could look into adding more advanced cycle detection, as long as we do it in a way that can be stripped out for production, i.e.

 - all work is done in `if (process.env.NODE_ENV !== 'production')` so that there is no performance impact in production.
 - It can only log/warn for errors, it must not reject any promises that would otherwise be fulfilled.

**Comment by amasad** [comment]
This sounds great. Cycles are incredibly hard to debug. Let me know how I can help
